auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -289,16 +325,48 @@ public TypeSpec createTypeAdapter(ClassName className, ClassName autoValueClassN
       }
     }
 
+    ProcessingEnvironment processingEnvironment = context.processingEnvironment();
+    TypeMirror typeAdapterFactory = processingEnvironment
+            .getElementUtils()
+            .getTypeElement("com.google.gson.TypeAdapterFactory")
+            .asType();
+    Types typeUtils = processingEnvironment.getTypeUtils();
+
     ImmutableMap<Property, FieldSpec> nonparameterizedAdapters = filterParameterizedAdapters(adapters, false);
     for (Map.Entry<Property, FieldSpec> entry : nonparameterizedAdapters.entrySet()) {
       Property prop = entry.getKey();
       FieldSpec field = entry.getValue();
-      if (entry.getKey().type instanceof ParameterizedTypeName) {
-        constructor.addStatement("this.$N = $N.getAdapter($L)", field, gsonParam,
-            makeType((ParameterizedTypeName) prop.type));
+      if (prop.typeAdapter != null) {
+        if (typeUtils.isAssignable(prop.typeAdapter, typeAdapterFactory)) {
+          constructor.addStatement("this.$N = new $T().create($N, $L)", field, TypeName.get(prop.typeAdapter),
+                  gsonParam, makeType(prop.type));
+        } else {
```

Comment:
Should we verify that the typeAdapter is actually a `TypeAdapter` and abort the processing if it isn't?